File: services/menos/menos/services/docling.py
"""Docling HTTP client for web content extraction."""

from dataclasses import dataclass

import httpx
from fastapi import HTTPException


@dataclass
class DoclingResult:
    """Extracted markdown content from Docling."""

    markdown: str
    title: str | None = None


class DoclingClient:
    """Client for Docling source conversion API."""

    def __init__(self, base_url: str, timeout: float = 30.0):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout

    async def extract_markdown(self, url: str) -> DoclingResult:
        """Extract markdown from a source URL via Docling."""
        payload = {
            "sources": [{"kind": "http", "url": url}],
            "options": {"to_formats": ["md"], "image_export_mode": "placeholder"},
        }

        endpoint = f"{self.base_url}/v1/convert/source"

        try:
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                response = await client.post(endpoint, json=payload)
            response.raise_for_status()
            data = response.json()
        except (httpx.HTTPError, ValueError) as exc:
            raise HTTPException(status_code=503, detail="Docling service unavailable") from exc

        markdown = _extract_markdown(data)
        if not markdown:
            raise HTTPException(status_code=503, detail="Docling returned no markdown")

        title = _extract_title(data) or _extract_title_from_markdown(markdown)
        return DoclingResult(markdown=markdown, title=title)


_MARKDOWN_KEYS = ("markdown", "md", "md_content")
_NESTED_KEYS = ("document", "documents", "output", "outputs", "data")
_TITLE_NESTED_KEYS = ("result",) + _NESTED_KEYS


def _extract_markdown_from_dict(data: dict) -> str | None:
    for key in _MARKDOWN_KEYS:
        val = data.get(key)
        if isinstance(val, str) and val.strip():
            return val
    for key in ("result",) + _NESTED_KEYS:
        if key in data:
            nested = _extract_markdown(data[key])
            if nested:
                return nested
    return None


def _extract_markdown(data: object) -> str | None:
    if isinstance(data, str):
        return data
    if isinstance(data, list):
        return next((m for item in data if (m := _extract_markdown(item))), None)
    if isinstance(data, dict):
        return _extract_markdown_from_dict(data)
    return None


def _str_if_nonempty(val: object) -> str | None:
    return val.strip() if isinstance(val, str) and val.strip() else None


def _extract_title_from_dict(data: dict) -> str | None:
    if result := _str_if_nonempty(data.get("title")):
        return result
    meta = data.get("metadata")
    if isinstance(meta, dict):
        if result := _str_if_nonempty(meta.get("title")):
            return result
    return next(
        (
            nested
            for key in _TITLE_NESTED_KEYS
            if key in data and (nested := _extract_title(data[key]))
        ),
        None,
    )


def _extract_title(data: object) -> str | None:
    if isinstance(data, list):
        return next((t for item in data if (t := _extract_title(item))), None)
    if isinstance(data, dict):
        return _extract_title_from_dict(data)
    return None


def _extract_title_from_markdown(markdown: str) -> str | None:
    for line in markdown.splitlines():
        stripped = line.strip()
        if stripped.startswith("#"):
            return stripped.lstrip("#").strip() or None
    return None

File: services/core/src/vault/docling.ts
import { HttpError } from "./errors";

export type DoclingResult = {
  markdown: string;
  title: string | null;
};

export type DoclingFetcher = (url: string, init?: RequestInit) => Promise<Response>;

const MARKDOWN_KEYS = ["markdown", "md", "md_content"] as const;
const NESTED_KEYS = ["document", "documents", "output", "outputs", "data"] as const;
const TITLE_NESTED_KEYS = ["result", ...NESTED_KEYS] as const;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function nonemptyString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim() !== "" ? value.trim() : undefined;
}

export function extractMarkdown(data: unknown): string | undefined {
  if (typeof data === "string") return data;
  if (Array.isArray(data)) {
    for (const item of data) {
      const markdown = extractMarkdown(item);
      if (markdown !== undefined) return markdown;
    }
    return undefined;
  }
  if (!isRecord(data)) return undefined;
  for (const key of MARKDOWN_KEYS) {
    const markdown = nonemptyString(data[key]);
    if (markdown !== undefined) return markdown;
  }
  for (const key of ["result", ...NESTED_KEYS]) {
    if (!(key in data)) continue;
    const markdown = extractMarkdown(data[key]);
    if (markdown !== undefined) return markdown;
  }
  return undefined;
}

export function extractTitle(data: unknown): string | undefined {
  if (Array.isArray(data)) {
    for (const item of data) {
      const title = extractTitle(item);
      if (title !== undefined) return title;
    }
    return undefined;
  }
  if (!isRecord(data)) return undefined;
  const direct = nonemptyString(data.title);
  if (direct !== undefined) return direct;
  if (isRecord(data.metadata)) {
    const metadataTitle = nonemptyString(data.metadata.title);
    if (metadataTitle !== undefined) return metadataTitle;
  }
  for (const key of TITLE_NESTED_KEYS) {
    if (!(key in data)) continue;
    const title = extractTitle(data[key]);
    if (title !== undefined) return title;
  }
  return undefined;
}

export function extractTitleFromMarkdown(markdown: string): string | undefined {
  for (const line of markdown.split(/\r?\n/)) {
    const stripped = line.trim();
    if (stripped.startsWith("#")) return stripped.replace(/^#+/, "").trim() || undefined;
  }
  return undefined;
}

function defaultFetcher(url: string, init?: RequestInit): Promise<Response> {
  return fetch(url, init);
}

export class DoclingClient {
  private readonly baseUrl: string;
  private readonly fetcher: DoclingFetcher;

  constructor(baseUrl: string, private readonly timeoutMs = 30_000, fetcher: DoclingFetcher = defaultFetcher) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.fetcher = fetcher;
  }

  async extractMarkdown(url: string): Promise<DoclingResult> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await this.fetcher(`${this.baseUrl}/v1/convert/source`, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({
          sources: [{ kind: "http", url }],
          options: { to_formats: ["md"], image_export_mode: "placeholder" },
        }),
        signal: controller.signal,
      });
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      const data: unknown = await response.json();
      const markdown = extractMarkdown(data);
      if (markdown === undefined || markdown === "") {
        throw new HttpError(503, "Docling returned no markdown");
      }
      return {
        markdown,
        title: extractTitle(data) ?? extractTitleFromMarkdown(markdown) ?? null,
      };
    } catch (error) {
      if (error instanceof HttpError) throw error;
      throw new HttpError(503, "Docling service unavailable");
    } finally {
      clearTimeout(timeout);
    }
  }
}
